File: src/handlers/actions.js
/**
 * Task action handlers.
 * Each function returns a human-friendly SMS reply string.
 */

const db = require('../db');
const { formatDueDate } = require('../utils/dates');

// ------------------------------------------------------------------
// CREATE
// ------------------------------------------------------------------

async function handleCreate(parsed, team, user) {
  const { task_title, owner_name, due_date } = parsed;

  if (!task_title) {
    return "What's the task? Try: \"Add task: write Q1 report\"";
  }

  // Resolve owner
  let owner = null;
  if (owner_name) {
    const matches = await db.findUserByNameInTeam(team.id, owner_name);
    if (matches.length === 1) {
      owner = matches[0];
    } else if (matches.length > 1) {
      return `Found multiple people named "${owner_name}". Can you be more specific?`;
    } else {
      return `I don't know anyone named "${owner_name}" on your team. Have them text this number first to join.`;
    }
  }

  const task = await db.createTask({
    teamId: team.id,
    ownerId: owner ? owner.id : user.id,
    createdById: user.id,
    title: task_title,
    dueDate: due_date || null,
  });

  const ownerName = owner ? owner.name : user.name;
  const duePart = due_date ? `, due ${formatDueDate(due_date)}` : '';
  return `Done — "${task_title}" assigned to ${ownerName}${duePart}.`;
}

// ------------------------------------------------------------------
// UPDATE STATUS
// ------------------------------------------------------------------

async function handleUpdate(parsed, team, user) {
  const { task_title, new_status, due_date } = parsed;

  if (!task_title) {
    return "Which task? Try: \"Mark homepage copy as done\"";
  }

  const tasks = await db.searchTasks(team.id, task_title);
  if (tasks.length === 0) {
    return `I couldn't find a task matching "${task_title}". Try again with a few keywords.`;
  }
  if (tasks.length > 1) {
    const list = tasks.slice(0, 3).map((t, i) => `${i + 1}. ${t.title}`).join('\n');
    return `Found multiple tasks:\n${list}\n\nBe more specific.`;
  }

  const task = tasks[0];
  const updates = {};
  if (new_status) updates.status = new_status;
  if (due_date) updates.due_date = due_date;

  if (Object.keys(updates).length === 0) {
    return "What should I update? You can change the status (done/blocked/open) or due date.";
  }

  await db.updateTask(task.id, team.id, updates);

  const statusEmoji = { done: '✓', blocked: '⚠', open: '○' }[new_status] || '';
  if (new_status === 'done') {
    return `${statusEmoji} Marked "${task.title}" as done. Nice work!`;
  } else if (new_status === 'blocked') {
    return `${statusEmoji} "${task.title}" marked as blocked. Loop in your team?`;
  } else if (due_date) {
    return `Updated — "${task.title}" now due ${formatDueDate(due_date)}.`;
  }
  return `Updated "${task.title}".`;
}

// ------------------------------------------------------------------
// ASSIGN
// ------------------------------------------------------------------

async function handleAssign(parsed, team, user) {
  const { task_title, owner_name } = parsed;

  if (!task_title) return "Which task do you want to reassign?";
  if (!owner_name) return "Who do you want to assign it to?";

  const tasks = await db.searchTasks(team.id, task_title);
  if (tasks.length === 0) return `Can't find a task matching "${task_title}".`;
  if (tasks.length > 1) {
    const list = tasks.slice(0, 3).map((t, i) => `${i + 1}. ${t.title}`).join('\n');
    return `Multiple matches:\n${list}\n\nBe more specific.`;
  }

  const matches = await db.findUserByNameInTeam(team.id, owner_name);
  if (matches.length === 0) return `"${owner_name}" isn't on your team yet.`;
  if (matches.length > 1) return `Multiple people named "${owner_name}". Who exactly?`;

  const task = tasks[0];
  const newOwner = matches[0];
  await db.updateTask(task.id, team.id, { owner_id: newOwner.id });

  return `Done — "${task.title}" reassigned to ${newOwner.name}.`;
}

// ------------------------------------------------------------------
// QUERY
// ------------------------------------------------------------------

async function handleQuery(parsed, team, user) {
  const { query_target } = parsed;

  let targetUser = user;
  let label = 'Your';

  if (query_target && query_target !== 'me') {
    if (query_target === 'all') {
      return await handleQueryAll(team);
    }
    const matches = await db.findUserByNameInTeam(team.id, query_target);
    if (matches.length === 0) return `"${query_target}" isn't on your team.`;
    if (matches.length > 1) return `Multiple people named "${query_target}". Be more specific.`;
    targetUser = matches[0];
    label = `${targetUser.name}'s`;
  }

  const tasks = await db.getOpenTasksForUser(targetUser.id);
  if (tasks.length === 0) return `${label} task list is empty — all clear!`;

  const lines = tasks.map(t => {
    const due = t.due_date ? ` (${formatDueDate(t.due_date)})` : '';
    const status = t.status === 'blocked' ? ' ⚠' : '';
    return `• ${t.title}${due}${status}`;
  });

  return `${label} open tasks:\n${lines.join('\n')}`;
}

async function handleQueryAll(team) {
  const tasks = await db.getTeamTasks(team.id, 'open');
  if (tasks.length === 0) return 'No open tasks — the team is all clear!';

  // Group by owner
  const byOwner = {};
  for (const t of tasks) {
    const name = t.owner?.name || 'Unassigned';
    if (!byOwner[name]) byOwner[name] = [];
    byOwner[name].push(t);
  }

  const sections = Object.entries(byOwner).map(([name, ts]) => {
    const lines = ts.map(t => {
      const due = t.due_date ? ` (${formatDueDate(t.due_date)})` : '';
      return `  • ${t.title}${due}`;
    });
    return `${name}:\n${lines.join('\n')}`;
  });

  return `Team tasks:\n\n${sections.join('\n\n')}`;
}

module.exports = { handleCreate, handleUpdate, handleAssign, handleQuery };

File: core/handlers/actions.py
"""
Task action handlers — each returns a human-friendly SMS reply string.
"""
import logging

from core.models import Task, User
from core.utils import format_due_date

logger = logging.getLogger(__name__)


# ------------------------------------------------------------------
# CREATE
# ------------------------------------------------------------------

def handle_create(parsed, team, user):
    task_title = parsed.get('task_title')
    owner_name = parsed.get('owner_name')
    due_date = parsed.get('due_date')

    if not task_title:
        return 'What\'s the task? Try: "Add task: write Q1 report"'

    owner = None
    if owner_name:
        matches = list(User.objects.filter(team=team, name__icontains=owner_name))
        if len(matches) == 1:
            owner = matches[0]
        elif len(matches) > 1:
            return f'Found multiple people named "{owner_name}". Can you be more specific?'
        else:
            return (
                f'I don\'t know anyone named "{owner_name}" on your team. '
                f'Have them text this number first to join.'
            )

    Task.objects.create(
        team=team,
        owner=owner or user,
        created_by=user,
        title=task_title,
        due_date=due_date or None,
        status='open',
    )

    owner_display = owner.name if owner else user.name
    due_part = f', due {format_due_date(due_date)}' if due_date else ''
    return f'Done \u2014 "{task_title}" assigned to {owner_display}{due_part}.'


# ------------------------------------------------------------------
# UPDATE STATUS / DUE DATE
# ------------------------------------------------------------------

def handle_update(parsed, team, user):
    task_title = parsed.get('task_title')
    new_status = parsed.get('new_status')
    due_date = parsed.get('due_date')

    if not task_title:
        return 'Which task? Try: "Mark homepage copy as done"'

    tasks = list(
        Task.objects.filter(team=team, title__icontains=task_title)
        .exclude(status='done')[:5]
    )
    if not tasks:
        return f'I couldn\'t find a task matching "{task_title}". Try again with a few keywords.'
    if len(tasks) > 1:
        items = '\n'.join(f'{i + 1}. {t.title}' for i, t in enumerate(tasks[:3]))
        return f'Found multiple tasks:\n{items}\n\nBe more specific.'

    task = tasks[0]
    if not new_status and not due_date:
        return 'What should I update? You can change the status (done/blocked/open) or due date.'

    if new_status:
        task.status = new_status
    if due_date:
        task.due_date = due_date
    task.save()

    status_emoji = {'done': '\u2713', 'blocked': '\u26a0', 'open': '\u25cb'}.get(new_status, '')
    if new_status == 'done':
        return f'{status_emoji} Marked "{task.title}" as done. Nice work!'
    if new_status == 'blocked':
        return f'{status_emoji} "{task.title}" marked as blocked. Loop in your team?'
    if due_date:
        return f'Updated \u2014 "{task.title}" now due {format_due_date(due_date)}.'
    return f'Updated "{task.title}".'


# ------------------------------------------------------------------
# ASSIGN
# ------------------------------------------------------------------

def handle_assign(parsed, team, user):
    task_title = parsed.get('task_title')
    owner_name = parsed.get('owner_name')

    if not task_title:
        return 'Which task do you want to reassign?'
    if not owner_name:
        return 'Who do you want to assign it to?'

    tasks = list(
        Task.objects.filter(team=team, title__icontains=task_title)
        .exclude(status='done')[:5]
    )
    if not tasks:
        return f'Can\'t find a task matching "{task_title}".'
    if len(tasks) > 1:
        items = '\n'.join(f'{i + 1}. {t.title}' for i, t in enumerate(tasks[:3]))
        return f'Multiple matches:\n{items}\n\nBe more specific.'

    matches = list(User.objects.filter(team=team, name__icontains=owner_name))
    if not matches:
        return f'"{owner_name}" isn\'t on your team yet.'
    if len(matches) > 1:
        return f'Multiple people named "{owner_name}". Who exactly?'

    task = tasks[0]
    new_owner = matches[0]
    task.owner = new_owner
    task.save()
    return f'Done \u2014 "{task.title}" reassigned to {new_owner.name}.'


# ------------------------------------------------------------------
# QUERY
# ------------------------------------------------------------------

def handle_query(parsed, team, user):
    query_target = parsed.get('query_target')

    target_user = user
    label = 'Your'

    if query_target and query_target != 'me':
        if query_target == 'all':
            return _handle_query_all(team)
        matches = list(User.objects.filter(team=team, name__icontains=query_target))
        if not matches:
            return f'"{query_target}" isn\'t on your team.'
        if len(matches) > 1:
            return f'Multiple people named "{query_target}". Be more specific.'
        target_user = matches[0]
        label = f"{target_user.name}'s"

    tasks = list(Task.objects.filter(owner=target_user, status='open').order_by('due_date'))
    if not tasks:
        return f'{label} task list is empty \u2014 all clear!'

    lines = []
    for t in tasks:
        due = f' ({format_due_date(str(t.due_date))})' if t.due_date else ''
        flag = ' \u26a0' if t.status == 'blocked' else ''
        lines.append(f'\u2022 {t.title}{due}{flag}')

    return f'{label} open tasks:\n' + '\n'.join(lines)


def _handle_query_all(team):
    tasks = list(
        Task.objects.filter(team=team, status='open')
        .select_related('owner')
        .order_by('due_date')
    )
    if not tasks:
        return 'No open tasks \u2014 the team is all clear!'

    by_owner = {}
    for t in tasks:
        name = t.owner.name if t.owner else 'Unassigned'
        by_owner.setdefault(name, []).append(t)

    sections = []
    for name, ts in by_owner.items():
        lines = []
        for t in ts:
            due = f' ({format_due_date(str(t.due_date))})' if t.due_date else ''
            lines.append(f'  \u2022 {t.title}{due}')
        sections.append(f'{name}:\n' + '\n'.join(lines))

    return 'Team tasks:\n\n' + '\n\n'.join(sections)
